Provide a mechanism for NSAttributedString to extend file access to WebContent process

brentfulgham opened the pull request:
#### 3c74d31715ca854fa0e02299d79c5756144d0d84
<pre>
Provide a mechanism for NSAttributedString to extend file access to WebContent process
<a href="https://bugs.webkit.org/show_bug.cgi?id=245538">https://bugs.webkit.org/show_bug.cgi?id=245538</a>
&lt;rdar://98333507&gt;

Reviewed by David Kilzer.

NSAttributedString in recent Cocoa OS releases makes use of the modern WebKit architecture,
and renders HTML content in a separate WebContent process from the main application. This
security improvement has created problems when an author attempts to create an NSAttributedString
from an HTML String that includes references to files in an application&apos;s bundle.

We need a way for WebKit&apos;s NSAttributedString extensions to pass file permission to the WebContent
process, otherwise the strings cannot be properly rendered.

Tests: New API test.

* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
(+[_WKAttributedStringWebViewCache readOnlyAccessPaths]): Added.
(+[_WKAttributedStringWebViewCache configuration]): Update to incorporate access paths
if relevant.
(+[_WKAttributedStringWebViewCache maybeConsumeBundlePaths:]): Added.
(+[NSAttributedString _loadFromHTMLWithOptions:contentLoader:completionHandler:]): Inititalize
access paths if relevant.
* Source/WebKit/UIProcess/API/Cocoa/NSAttributedStringPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdditionalReadAccessAllowedURLs.mm:
(TEST): Added.

Canonical link: <a href="https://commits.webkit.org/254968@main">https://commits.webkit.org/254968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22cf55e23894f4cd21e0fac10e0a232785114253

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90722 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35302 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21314 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100034 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158354 "Failed to checkout and rebase branch from PR 4608") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94731 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33801 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28926 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83079 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96427 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96377 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26923 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77542 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26741 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81683 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81445 "Found 2 new API test failures: TestWebKitAPI.WebKit.NSAttributedStringWithReadOnlyPaths, TestWebKitAPI.WebKit.NSAttributedStringWithAndWithoutReadOnlyPaths (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69772 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34891 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15482 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32702 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16462 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36468 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39402 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1515 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38391 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35551 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -133,12 +136,27 @@ @implementation _WKAttributedStringWebViewCache
 
++ (NSMutableArray<NSURL *> *)readOnlyAccessPaths
+{
+    static auto* readOnlyAccessPaths = [[NSMutableArray alloc] initWithCapacity:maximumReadOnlyAccessPaths];
```
This should at least use a `RetainPtr<NSMutableArray>`, and maybe a `NeverDestroyed<RetainPtr<NSMutableArray>>`.

It should not be a raw pointer.

brentfulgham replied:
Will fix.

brentfulgham replied:
Done!

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -45,11 +45,13 @@
 #endif
 
 NSString * const NSReadAccessURLDocumentOption = @"ReadAccessURL";
+NSString * const WKReadAccessFileURLsOption = @"_WKReadAccessFileURLsOption";
```
I feel like the constant name should also be prefixed by an underscore (`_WKReadAccessFileURLsOption`) since it's not API yet (and to match the value), but I'm not sure what the convention is.

I also see a style like this used in some places:
```
NSString * const _WKReadAccessFileURLsOption = @"WKReadAccessFileURLsOption";
```

The latter style documents that the variable is not public API, but uses (what would be) a public API key name in case it becomes public later.

Please check with someone who does regular API review to find out the correct way to define these before landing.

brentfulgham replied:
Since this is not a persistent key (it's not stored in User Defaults, for example), I think it's safe for us to use the underscored version of the string until an explicit decision to make it API.

I'll revise the name of the constant to use a prefix, too, to signify that these are currently SPI. I'll also file a radar to make it API for the next SDK update.

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -133,12 +136,27 @@ @implementation _WKAttributedStringWebViewCache
-        [configuration setProcessPool:adoptNS([[WKProcessPool alloc] init]).get()];
+
+        RetainPtr<WKProcessPool> processPool;
+        if (self.readOnlyAccessPaths.count) {
```
Should check for the maximum count here (`maximumReadOnlyAccessPaths`) as well before setting it.

brentfulgham replied:
We shouldn't ever be in a state where this is above 2, since we guard against it when ready the value from @"WKReadAccessFileURLsOption". I'll add a RELEASE_ASSERT here, since a > 2 value would indicate some trickery that should probably be a catastrophic app termination.

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +182,25 @@ + (void)clearConfiguration
 
++ (void)maybeConsumeBundlePaths:(NSDictionary<NSAttributedStringDocumentReadingOptionKey, id> *)options
+{
+    auto* readAccessPaths = dynamic_objc_cast<NSArray<NSURL *>>(options[WKReadAccessFileURLsOption]);
```
You should know that the `dynamic_objc_cast<>` here only checks that the outer object is an `NSArray` (or `NSMutableAray`).

It does not check that the contents of the `NSArray` are only `NSURL` objects.

Objective-C generics are merely syntactic sugar to catch mistakes during compiling, but are never type-checked at runtime.

brentfulgham replied:
Ah, okay. I think it is sufficient to validate on setting the value (as you propose below).

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +182,25 @@ + (void)clearConfiguration
+    if ([readAccessPaths isEqualToArray:self.readOnlyAccessPaths])
+        return;
+
+    [self.readOnlyAccessPaths setArray:readAccessPaths];
```
If you want to verify that `readAccessPaths` only contains `NSURL` objects (and it seems you do), you should write a `for()` loop to iterate through the objects here to check for that _before_ this code.  Something like this:

```
    for (id url in readOnlyAccessPaths) {
        if (!dynamic_objc_cast<NSURL>(url))
            return;
    }
```

brentfulgham replied:
Will do.

ddkilzer commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/AdditionalReadAccessAllowedURLs.mm:
```diff
@@ -33,13 +33,32 @@
 #import <WebKit/_WKRemoteObjectRegistry.h>
 #import <wtf/RetainPtr.h>
 
+static std::pair<NSURL *, NSURL *> readableAndUnreadableDirectories()
```
The `std::pair<>` return values are bare Objective-C pointers, so some times when the test runs, they'll probably cause crashes.

The return type should probably be:  `std::pair<RetainPtr<NSURL>, RetainPtr<NSURL>>`

darinadler replied:
It is not true that we we need `RetainPtr` here. Autorelease works just as well on the pair as it would work if the result was just an `NSURL`.

We can and should start using ARC when we write test code, making it much easier to get this right and avoiding any debate about this.

brentfulgham replied:
Fixed!

ddkilzer commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/AdditionalReadAccessAllowedURLs.mm:
```diff
@@ -105,4 +113,83 @@
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability-new"
+    NSMutableDictionary *options = [[NSMutableDictionary alloc] initWithObjectsAndKeys:@[ readableDirectoryURL, unreadableDirectoryURL, bundlePathURL ], WKReadAccessFileURLsOption, nil];
```
The `NSMutableDictionary` is leaked here because you don't call `[options release]` later.  (I assume this is _not_ compiled with ARC.)

Use a `RetainPtr< NSMutableDictionary >` and `adoptNS()` to fix the leak.

brentfulgham replied:
Ugh. Fixed!

ddkilzer commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/AdditionalReadAccessAllowedURLs.mm:
```diff
@@ -105,4 +113,83 @@
+    
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability-new"
+    NSMutableDictionary *options = [[NSMutableDictionary alloc] initWithObjectsAndKeys:@[ readableDirectoryURL ], WKReadAccessFileURLsOption, nil];
```
Ditto for leaking `options` here.

brentfulgham replied:
Fixed.

ddkilzer reviewed (approved):
r=me if you resolve the comments.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +182,25 @@ + (void)clearConfiguration
+        return;
+
+    if (!readAccessPaths.count)
+        return;
```
Same question. But why add a special case at all? Does it do harm to accept an empty array?

brentfulgham replied:
It's not harmful -- it just return to the default state of no additional directories being sent to the WebContent process. I'll revise to handle that, and add a test case.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +182,25 @@ + (void)clearConfiguration
+{
+    auto* readAccessPaths = dynamic_objc_cast<NSArray<NSURL *>>(options[WKReadAccessFileURLsOption]);
+    if (!readAccessPaths)
+        return;
```
Why is it correct to leave the old paths set when this is not specified? If a new path was specified, we’d overwrite the old ones, but unspecified just leaves the old ones alone? I think any time this is called, we should reset the old ones. I also think the method name should make the clear.

Or we can say it’s not legal to call this multiple times, in which case I think we should check that and not try to handle overwriting an old set with a new one.

brentfulgham replied:
I didn't really consider that scenario, but you are correct -- it's a reasonable thing to do, and probably safer. One could imagine a case where someone needs to render an HTML string that requires access to some place we don't usually need access. Once that operation is compete, there is no need to continue using the cached WKWebView that happens to have that permission. Passing an empty set of directories should just tear down the old WKWebView and generate a "default" view that doesn't have those permissions.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +183,30 @@ + (void)clearConfiguration
+
+    for (id fileURL in readAccessFileURLs) {
+        if (!dynamic_objc_cast<NSURL>(fileURL))
+            return;
```
We throw an exception when the count is too high. What other types of errors should be throwing exceptions rather than silently doing nothing? Also, maybe we should always clear the old paths even if we do throw an exception?

brentfulgham replied:
It seems reasonable to throw if we are given incorrect types (e.g., NSString* paths instead of NSURL*).

We can also throw if the NSURLs are for non-file URLS (which are the only types this API is meant to support).

Clearing the cached state if incorrect arguments are passed seems like a good idea. I'll add that and a test case.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -133,12 +137,28 @@ @implementation _WKAttributedStringWebViewCache
     return configuration;
 }
 
++ (RetainPtr<NSMutableArray<NSURL *>>)readOnlyAccessPaths
```
I think should just return a pointer, not a `RetainPtr`. It’s not transferring ownership.

I think this should just be a function, not a method.

brentfulgham replied:
Okay, I'll switch back to pointer and make this a static function.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +184,45 @@ + (void)clearConfiguration
+    auto* url = dynamic_objc_cast<NSURL>(maybeFileURL);
+    if (!url)
+        errorMessage = @"The NSDictionary associated with _WKReadAccessFileURLsOption may only contain NSURL objects.";
+    else if (!url.fileURL)
```
Isn’t this property named `isFileURL`? I’m surprised that `fileURL` exists.

brentfulgham replied:
It's strange that it works -- I can see that the getter is declared "isFileURL", but this compiled. Strange -- I'll fix this. The test case properly threw with a non-FileURL, so I didn't notice.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +184,45 @@ + (void)clearConfiguration
+        return;
+
+    if (self.readOnlyAccessPaths.get().count) {
+        [self.readOnlyAccessPaths removeAllObjects];
```
Does this work, or does it just remove the objects from a copy of the array? I think this might need to be `[self.readOnlyAccessPaths setArray:@[]]`.

brentfulgham replied:
I think this is correct as-is, since `readOnlyAccessPaths` returns a pointer (prior to your edits, a RetainPtr) to the NeverDestroyed NSMutableArray. I'll confirm in the debugger, since the test cases don't exercise this well (if we have no required directories, we can't tell if we have a set of unused directories in the `NSMutableArray` or a truly empty array!)

darinadler replied:
Oh, OK, that’s the function that should be a function and not a method! I didn’t realize that.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +184,45 @@ + (void)clearConfiguration
+    if (readAccessFileURLs)
+        [self.readOnlyAccessPaths setArray:readAccessFileURLs];
+    else
+        [self.readOnlyAccessPaths removeAllObjects];
```
Does this work, or does it just remove the objects from a copy of the array? I think this might need to be `[self.readOnlyAccessPaths setArray:@[]]`.

brentfulgham replied:
Ditto the above. I will confirm.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +184,45 @@ + (void)clearConfiguration
+{
+    auto* readAccessFileURLs = dynamic_objc_cast<NSArray<NSURL *>>(options[_WKReadAccessFileURLsOption]);
+    if (readAccessFileURLs.count > maximumReadOnlyAccessPaths)
+        [NSException raise:NSInvalidArgumentException format:@"_WKReadAccessFileURLsOption may have at most two additional directories."];
```
I think we went to do the same thing here as where we raise the exception above in `validateEntry:` so they should share code.

darinadler replied:
The same thing being the "remove any items that are in the array from a previous call" code.

brentfulgham replied:
Will do.

darinadler commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -164,6 +184,45 @@ + (void)clearConfiguration
+
++ (void)maybeConsumeBundlePaths:(NSDictionary<NSAttributedStringDocumentReadingOptionKey, id> *)options
+{
+    auto* readAccessFileURLs = dynamic_objc_cast<NSArray<NSURL *>>(options[_WKReadAccessFileURLsOption]);
```
We should throw an exception if it’s not an `NSArray` if we’re throwing exceptions when elements of the array are the wrong type.

darinadler replied:
And doing the "remove item from a previous call" code.

brentfulgham replied:
Will do.

brentfulgham committed:
Provide a mechanism for NSAttributedString to extend file access to WebContent process
https://bugs.webkit.org/show_bug.cgi?id=245538
<rdar://98333507>

Reviewed by David Kilzer.

NSAttributedString in recent Cocoa OS releases makes use of the modern WebKit architecture,
and renders HTML content in a separate WebContent process from the main application. This
security improvement has created problems when an author attempts to create an NSAttributedString
from an HTML String that includes references to files in an application's bundle.

We need a way for WebKit's NSAttributedString extensions to pass file permission to the WebContent
process, otherwise the strings cannot be properly rendered.

Tests: New API test.

* Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
(+[_WKAttributedStringWebViewCache readOnlyAccessPaths]): Added.
(+[_WKAttributedStringWebViewCache configuration]): Update to incorporate access paths
if relevant.
(+[_WKAttributedStringWebViewCache maybeConsumeBundlePaths:]): Added.
(+[NSAttributedString _loadFromHTMLWithOptions:contentLoader:completionHandler:]): Inititalize
access paths if relevant.
* Source/WebKit/UIProcess/API/Cocoa/NSAttributedStringPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdditionalReadAccessAllowedURLs.mm:
(TEST): Added.

Canonical link: https://commits.webkit.org/254968@main
